python/strings_udf/cpp/include/cudf/strings/udf/udf_string.hpp

```diff
@@ -0,0 +1,532 @@
+   *            or all compared characters match but `bytes < size_bytes()`.
+   *         >0 Either the value of the first character of this string that does
+   *            not match is ordered after the corresponding character in `str`,
+   *            or all compared characters match but `bytes > size_bytes()`.
+   */
+  __device__ int compare(char const* str, cudf::size_type bytes) const;
+
+  /**
+   * @brief Returns true if `rhs` matches this string exactly
+   */
+  __device__ bool operator==(cudf::string_view const rhs) const;
+
+  /**
+   * @brief Returns true if `rhs` does not match this string
+   */
+  __device__ bool operator!=(cudf::string_view const rhs) const;
+
+  /**
+   * @brief Returns true if this string is ordered before `rhs`
+   */
+  __device__ bool operator<(cudf::string_view const rhs) const;
+
+  /**
+   * @brief Returns true if `rhs` is ordered before this string
+   */
+  __device__ bool operator>(cudf::string_view const rhs) const;
+
+  /**
+   * @brief Returns true if this string matches or is ordered before `rhs`
+   */
+  __device__ bool operator<=(cudf::string_view const rhs) const;
+
+  /**
+   * @brief Returns true if `rhs` matches or is ordered before this string
+   */
+  __device__ bool operator>=(cudf::string_view const rhs) const;
+
+  /**
+   * @brief Remove all bytes from this string
+   *
+   * All pointers, references, and iterators are invalidated.
+   */
+  __device__ void clear();
+
+  /**
+   * @brief Resizes string to contain `count` bytes
+   *
+   * If `count > size_bytes()` then zero-padding is added.
+   * If `count < size_bytes()` then the string is truncated to size `count`.
+   *
+   * All pointers, references, and iterators may be invalidated.
+   *
+   * @param count Size in bytes of this string.
+   */
+  __device__ void resize(cudf::size_type count);
+
+  /**
+   * @brief Reserve `count` bytes in this string
+   *
+   * If `count > capacity()`, new memory is allocated and `capacity()` will
+   * be greater than or equal to `count`.
+   * There is no effect if `count <= capacity()`.
+   *
+   * @param count Total number of bytes to reserve for this string
+   */
+  __device__ void reserve(cudf::size_type count);
+
+  /**
+   * @brief Returns the number of bytes that the string has allocated
+   */
+  __device__ cudf::size_type capacity() const;
+
+  /**
+   * @brief Reduces internal allocation to just `size_bytes()`
+   *
+   * All pointers, references, and iterators may be invalidated.
+   */
+  __device__ void shrink_to_fit();
+
+  /**
+   * @brief Moves the contents of `str` into this string instance
+   *
+   * @param str String to move
+   * @return This string new contents
+   */
+  __device__ udf_string& assign(udf_string&& str);
+
+  /**
+   * @brief Replaces the contents of this string with contents of `str`
+   *
+   * @param str String to copy
+   * @return This string new contents
+   */
+  __device__ udf_string& assign(cudf::string_view const str);
```

Comment:
In this case I assume the contents of `str` are _not_ invalidated?

Reply:
Correct.